backend/infrahub/core/diff/merger/serializer.py

```diff
@@ -169,38 +172,105 @@ def _serialize_attribute(
                 property_diff=property_diff, python_value_type=python_type
             )
             for action, value in actions_and_values:
+                # we only delete attributes when the whole attribute is deleted
+                if action is DiffAction.REMOVED and attribute_diff.action is not DiffAction.REMOVED:
+                    continue
                 prop_dicts.append(
                     PropertyMergeDict(
                         property_type=property_diff.property_type.value,
                         action=self._to_action_str(action=action),
                         value=value,
                     )
                 )
-        attr_dict = AttributeMergeDict(
-            name=attribute_diff.name,
-            action=self._to_action_str(action=attribute_diff.action),
-        )
+        attr_dict = None
+        if attribute_diff.action in (DiffAction.ADDED, DiffAction.REMOVED):
+            attr_dict = AttributeMergeDict(
+                name=attribute_diff.name,
+                action=self._to_action_str(action=attribute_diff.action),
+            )
         attr_prop_dict = AttributePropertyMergeDict(
             node_uuid=node_uuid, attribute_name=attribute_diff.name, properties=prop_dicts
         )
         return attr_dict, attr_prop_dict
 
     def _serialize_relationship_element(
-        self, relationship_diff: EnrichedDiffSingleRelationship, relationship_identifier: str
+        self, relationship_diff: EnrichedDiffSingleRelationship, relationship_identifier: str, node_uuid: str
     ) -> tuple[list[RelationshipMergeDict], list[RelationshipPropertyMergeDict]]:
         relationship_dicts = []
+        # start with default values for IS_VISIBLE and IS_PROTECTED b/c we always want to update them during a merge
+        added_property_dicts: dict[DatabaseEdgeType, PropertyMergeDict] = {
```

Comment:
these are initialized with IS_PROTECTED and IS_VISIBLE starting values to make sure that they are always updated for a given relationship